src/assets/TokenDetectionController.ts

```diff
@@ -0,0 +1,243 @@
+      addTokens: TokensController['addTokens'];
+      getTokenListState: () => TokenListState;
+      getTokensState: () => TokensState;
+    },
+    config?: Partial<TokenDetectionConfig>,
+    state?: Partial<BaseState>,
+  ) {
+    super(config, state);
+    this.defaultConfig = {
+      interval: DEFAULT_INTERVAL,
+      networkType: MAINNET,
+      selectedAddress: '',
+      tokens: [],
+    };
+    this.initialize();
+    this.getTokensState = getTokensState;
+    this.getTokenListState = getTokenListState;
+    this.addTokens = addTokens;
+    onTokensStateChange(({ tokens }) => {
+      this.configure({ tokens });
+    });
+
+    onPreferencesStateChange(({ selectedAddress }) => {
+      const actualSelectedAddress = this.config.selectedAddress;
+      if (selectedAddress !== actualSelectedAddress) {
+        this.configure({ selectedAddress });
+        this.detectTokens();
+      }
+    });
+
+    onNetworkStateChange(({ provider }) => {
+      this.configure({ networkType: provider.type });
+    });
+    this.getBalancesInSingleCall = getBalancesInSingleCall;
+    this.start();
+  }
+
+  /**
+   * Start polling for the currency rate.
+   */
+  async start() {
+    if (!this.isMainnet() || this.disabled) {
+      return;
+    }
+
+    await this.startPolling();
+  }
+
+  /**
+   * Stop polling for the currency rate.
+   */
+  stop() {
+    this.stopPolling();
+  }
+
+  private stopPolling() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  /**
+   * Starts a new polling interval.
+   *
+   * @param interval - An interval on which to poll.
+   */
+  private async startPolling(interval?: number): Promise<void> {
+    interval && this.configure({ interval }, false, false);
+    this.stopPolling();
+    await this.detectTokens();
+    this.intervalId = setInterval(async () => {
+      await this.detectTokens();
+    }, this.config.interval);
+  }
+
+  /**
+   * Checks whether network is mainnet or not.
+   *
+   * @returns Whether current network is mainnet.
+   */
+  isMainnet = (): boolean => this.config.networkType === MAINNET;
+
+  /**
+   * Triggers asset ERC20 token auto detection for each contract address in contract metadata on mainnet.
+   */
+  async detectTokens() {
+    /* istanbul ignore if */
+    if (!this.isMainnet() || this.disabled) {
+      return;
+    }
+
+    const tokensAddresses = this.config.tokens.map(
+      /* istanbul ignore next*/ (token) => token.address.toLowerCase(),
+    );
+    const { tokenList } = this.getTokenListState();
+    const tokensToDetect: string[] = [];
+    for (const address in tokenList) {
+      if (!tokensAddresses.includes(address)) {
+        tokensToDetect.push(address);
+      }
+    }
+    const sliceOfTokensToDetect = [];
+    sliceOfTokensToDetect[0] = tokensToDetect.slice(0, 1000);
+    sliceOfTokensToDetect[1] = tokensToDetect.slice(
+      1000,
+      tokensToDetect.length - 1,
+    );
```

Comment:
Can we update this to the following code? This is the pattern we use in the extension detect-token.
    const sliceOfTokensToDetect = [
      tokensToDetect.slice(0, 1000),
      tokensToDetect.slice(1000, tokensToDetect.length - 1),
    ];

Reply:
This is a tiny semantic change and we can make the change in a followup PR to clean up code. We also will want to do an audit of difference between the extension token-detection controller and this one so we can handle that when we switch the extension to use of the controllers repo version.